libethereum/ChainParams.cpp

```diff
@@ -34,60 +34,110 @@ using namespace dev;
+	js::mObject genesis = val.get_obj();
+
+	cp.parentHash = h256(genesis["parentHash"].get_str());
+	cp.author = genesis.count("coinbase") ? h160(genesis["coinbase"].get_str()) : h160(genesis["author"].get_str());
+	cp.difficulty = genesis.count("difficulty") ? u256(fromBigEndian<u256>(fromHex(genesis["difficulty"].get_str()))) : 0;
+	cp.gasLimit = u256(fromBigEndian<u256>(fromHex(genesis["gasLimit"].get_str())));
+	cp.gasUsed = genesis.count("gasUsed") ? u256(fromBigEndian<u256>(fromHex(genesis["gasUsed"].get_str()))) : 0;
+	cp.timestamp = u256(fromBigEndian<u256>(fromHex(genesis["timestamp"].get_str())));
+	cp.extraData = bytes(fromHex(genesis["extraData"].get_str()));
+
+	// magic code for handling ethash stuff:
+	if ((genesis.count("mixhash") || genesis.count("mixHash")) && genesis.count("nonce"))
 	{
-		accountStartNonce = u256(fromBigEndian<u256>(fromHex(params["accountStartNonce"].get_str())));
-		maximumExtraDataSize = u256(fromBigEndian<u256>(fromHex(params["maximumExtraDataSize"].get_str())));
-		tieBreakingGas = params.count("tieBreakingGas") ? params["tieBreakingGas"].get_bool() : true;
-		blockReward = u256(fromBigEndian<u256>(fromHex(params["blockReward"].get_str())));
-		for (auto i: params)
-			if (i.first != "accountStartNonce" && i.first != "maximumExtraDataSize" && i.first != "blockReward" && i.first != "tieBreakingGas")
-				otherParams[i.first] = i.second.get_str();
+		h256 mixHash(genesis[genesis.count("mixhash") ? "mixhash" : "mixHash"].get_str());
+		h64 nonce(genesis["nonce"].get_str());
+		cp.sealFields = 2;
+		cp.sealRLP = rlp(mixHash) + rlp(nonce);
 	}
+	cp.stateRoot = _stateRoot ? _stateRoot : cp.calculateStateRoot();
+	return cp;
+}
+
+ChainParams ChainParams::setGenesisState(string const& _json, unordered_map<Address, PrecompiledContract> const& _precompiled, ChainParams const& _org) const
+{
+	ChainParams cp(_org);
+	cp.precompiled = _precompiled;
+	cp.genesisState = jsonToAccountMap(_json, nullptr, &cp.precompiled);
+	return cp;
+}
+
 
+ChainParams ChainParams::loadConfig(string const& _json, bool _importGenesis, h256 const& _stateRoot) const
+{
+	ChainParams cp = ChainParams(*this);
+	js::mValue val;
+	json_spirit::read_string(_json, val);
+	js::mObject obj = val.get_obj();
+
+	cp.sealEngineName = obj["sealEngine"].get_str();
+	// params
+	js::mObject params = obj["params"].get_obj();
+	cp.accountStartNonce = u256(fromBigEndian<u256>(fromHex(params["accountStartNonce"].get_str())));
+	cp.maximumExtraDataSize = u256(fromBigEndian<u256>(fromHex(params["maximumExtraDataSize"].get_str())));
+	cp.tieBreakingGas = params.count("tieBreakingGas") ? params["tieBreakingGas"].get_bool() : true;
+	cp.blockReward = u256(fromBigEndian<u256>(fromHex(params["blockReward"].get_str())));
+	for (auto i: params)
+		if (i.first != "accountStartNonce" && i.first != "maximumExtraDataSize" && i.first != "blockReward" && i.first != "tieBreakingGas")
+			cp.otherParams[i.first] = i.second.get_str();
 	// genesis
+	if (_importGenesis)
```

Comment:
I think it is better to import it and overwrite it when needed instead of having this boolean flag.